tools/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c

```diff
@@ -0,0 +1,38 @@
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "queue.h"
+
+/* FreeRTOS+TCP includes. */
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_IP_Private.h"
+#include "FreeRTOS_TCP_IP.h"
+
+#include "../../utility/memory_assignments.c"
+
+/* This proof assumes that pxDuplicateNetworkBufferWithDescriptor is implemented correctly. */
+
+void publicTCPReturnPacket( FreeRTOS_Socket_t *pxSocket, NetworkBufferDescriptor_t *pxNetworkBuffer, uint32_t ulLen, BaseType_t xReleaseAfterSend );
+
+/* Abstraction of pxDuplicateNetworkBufferWithDescriptor*/
+NetworkBufferDescriptor_t *pxDuplicateNetworkBufferWithDescriptor( NetworkBufferDescriptor_t * const pxNetworkBuffer,
+	BaseType_t xNewLength ) {
+	size_t bufferSize = sizeof(NetworkBufferDescriptor_t);
+	NetworkBufferDescriptor_t *pxNetworkBuffer = ensure_FreeRTOS_NetworkBuffer_is_allocated();
+	if (ensure_memory_is_valid(pxNetworkBuffer, bufferSize)) {
+		pxNetworkBuffer->pucEthernetBuffer = safeMalloc(sizeof(TCPPacket_t));
+	}
+	return pxNetworkBuffer;
+}
+
+void harness() {
+	size_t bufferSize = sizeof(NetworkBufferDescriptor_t);
+	FreeRTOS_Socket_t *pxSocket = ensure_FreeRTOS_Socket_t_is_allocated();
+	NetworkBufferDescriptor_t *pxNetworkBuffer = ensure_FreeRTOS_NetworkBuffer_is_allocated();
+	if (ensure_memory_is_valid(pxNetworkBuffer, bufferSize)) {
+		pxNetworkBuffer->pucEthernetBuffer = safeMalloc(sizeof(TCPPacket_t));
+	}
+	uint32_t ulLen;
+	BaseType_t xReleaseAfterSend;
+	__CPROVER_assume(pxSocket != NULL || pxNetworkBuffer != NULL); /* Both can not be NULL at the same time. */
```

Comment:
Is there a line of documentation you can point to for this?

Reply:
I have not seen any documentation. The code does not expect both these to be equal to `NULL` at the same time.

Reply:
This seems like it should be an assert in the code then. Implicit assumptions make me nervous